<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

## Connect to Hyperliquid API 
### setup function

In [0]:
#| echo: false
#| output: asis
show_doc(setup)

---

### setup

```python
def setup(
    base_url:NoneType=None, skip_ws:bool=False, perp_dexs:NoneType=None, config:str='../config_hyperliquid.json',
    debug:bool=False
):
```

*Call self as a function.*

### Example

In [ ]:
#| eval: false
address, info, exchange = setup(base_url=constants.MAINNET_API_URL, skip_ws=True)

## Get historical prerpetual price data
### retrive_hyperliquid_perp_price function

In [0]:
#| echo: false
#| output: asis
show_doc(retrieve_hyperliquid_perp_price)

---

### retrieve_hyperliquid_perp_price

```python
def retrieve_hyperliquid_perp_price(
    coin:str='ETH', interval:str='1h', end_date:str='2026-07-08T12:24:54Z', start_date:str='2026-07-06T12:24:54Z',
    info:NoneType=None
):
```

*Retrieves historical candle data from Hyperliquid for a given coin and time interval.*

Args:
    coin (str, optional): Coin symbol (e.g. "ETH"). Defaults to "ETH".
    interval (str, optional): Candle interval ("1m", "5m", "15m", "1h", "4h", "1d"). Defaults to "1h".
    end_date (str, optional): End datetime in ISO 8601 format. Defaults to current UTC time.
    start_date (str, optional): Start datetime in ISO 8601 format. Defaults to 2 days before end_date.
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.

Returns:
    pandas.DataFrame: DataFrame containing the OHLCV data with columns:
        - datetime: Timestamp for the candle (UTC)
        - open: Opening price of the interval
        - high: Highest traded price in the interval
        - low: Lowest traded price in the interval
        - close: Closing price of the interval
        - volume: Trading volume in the interval
        - coin: Coin symbol
    Returns None if the API request fails or returns no data.

Notes:
    - All datetime values are in UTC timezone
    - Requires Hyperliquid Info client to be initialized

### Example

You will need a api key and secret key from the Hyperliquid API. Store this into the `config_hyperliquid.json` file in the same directory as your script. See: https://app.hyperliquid.xyz/API

The best is to first call the `setup` function once to initialize the Hyperliquid Info client. Then pass "info" to any function that requires it. That will avoid calling the `setup` function multiple times.

Typical usage:

In [ ]:
#| eval: false
perp = retrieve_hyperliquid_perp_price(coin="ETH", interval="1h",info=info)
print(perp.head())

             datetime    open    high     low   close      volume coin
0 2025-12-03 00:00:00  2994.7  3007.5  2984.3  3001.9  16407.3455  ETH
1 2025-12-03 01:00:00  3002.3  3026.6  2998.1  3014.7   7931.4838  ETH
2 2025-12-03 02:00:00  3014.6  3046.4  3013.3  3030.1  16840.5078  ETH
3 2025-12-03 03:00:00  3029.6  3037.0  3019.8  3029.6  13628.3697  ETH
4 2025-12-03 04:00:00  3029.6  3058.8  3028.4  3055.4  15111.4473  ETH


## List of spot tickers
### spot_tickers function

In [0]:
#| echo: false
#| output: asis
show_doc(spot_tickers)

---

### spot_tickers

```python
def spot_tickers(
    coin:str='ETH', base:str='USDC', info:NoneType=None
):
```

*Retrieves current tickers for a given coin.*

Args:
    coin (str, optional): Coin symbol (e.g. "ETH"). Defaults to "ETH".
    base (str, optional): Coin symbol (e.g. "USDC"). Defaults to "USDC".
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.

Returns:
    spot ticker (non intuitive symbol)
    Returns None if the API request fails or returns no data.

Notes:
    - Requires Hyperliquid Info client to be initialized

### Example

In [ ]:
#| eval: false
stk = spot_tickers(info=info,coin="ETH",base="USDC")
print(f'spot ticker for ETH: ', stk)

spot ticker for ETH:  @151


## Get historical spot price
### retrieve_hyperliquid_spot_price function

In [0]:
#| echo: false
#| output: asis
show_doc(retrieve_hyperliquid_spot_price)

---

### retrieve_hyperliquid_spot_price

```python
def retrieve_hyperliquid_spot_price(
    coin:str='ETH', base:str='USDC', interval:str='1h', end_date:str='2026-07-08T12:24:54Z',
    start_date:str='2026-07-06T12:24:54Z', info:NoneType=None, recheck:bool=False
):
```

*Call self as a function.*

### Example

The ticker for Ethereum (ETH) is "@151" ("UETH"). There are other unusual choices but the function handles these choices by changing the ticker inside the function. 

In [ ]:
#| eval: false
spot=retrieve_hyperliquid_spot_price(info=info)
print(spot.tail())

              datetime    open    high     low   close    volume coin
44 2025-12-04 20:00:00  3120.6  3145.6  3106.6  3143.1  202.4952  ETH
45 2025-12-04 21:00:00  3141.0  3154.2  3120.0  3123.9  339.5227  ETH
46 2025-12-04 22:00:00  3125.1  3148.6  3124.4  3144.5   44.5737  ETH
47 2025-12-04 23:00:00  3141.9  3144.7  3130.9  3133.9   54.7757  ETH
48 2025-12-05 00:00:00  3133.9  3156.3  3133.4  3147.9   35.9388  ETH


If you want to find out which ticker is used for a given coin, you can use the `spot_tickers` function. For example, for Ethereum (ETH) quoted in USDC is:

In [ ]:
#| eval: false
stk = spot_tickers(info=info,coin="ETH",base="USDC")
print(f'spot ticker for ETH: ', stk)

spot ticker for ETH:  @151


## List all tokens in Hyperliquid
### hyperliquid_tokens function

In [0]:
#| echo: false
#| output: asis
show_doc(hyperliquid_tokens)

---

### hyperliquid_tokens

```python
def hyperliquid_tokens(
    info:NoneType=None, rm_delisted:bool=True
):
```

*Call self as a function.*

### Example

In [ ]:
#| eval: false
tokens = hyperliquid_tokens(info)
print(tokens)

     szDecimals  name  maxLeverage  marginTableId  isDelisted  onlyIsolated  \
0             5   BTC           40             56       False         False   
1             4   ETH           25             55       False         False   
2             2  ATOM            5              5       False         False   
4             1  DYDX            5              5       False         False   
5             2   SOL           20             54       False         False   
..          ...   ...          ...            ...         ...           ...   
215           0   MON            5              5       False         False   
216           0   MET            3              3       False         False   
218           0    CC            3              3       False         False   
219           1   ICP            5              5       False         False   
220           0  AERO            3              3       False         False   

    marginMode  
0          NaN  
1          NaN  


## Get funding rate history
### retrieve_hyperliquid_funding_history function

In [0]:
#| echo: false
#| output: asis
show_doc(funding_calc)

---

### funding_calc

```python
def funding_calc(
    rate, premium, max_rate:float=0.0005, min_rate:float=-0.0005
):
```

*Call self as a function.*

In [0]:
#| echo: false
#| output: asis
show_doc(retrieve_hyperliquid_funding_history)

---

### retrieve_hyperliquid_funding_history

```python
def retrieve_hyperliquid_funding_history(
    coin:str='ETH', end_date:str='2026-07-08T12:24:54Z', start_date:str='2026-07-06T12:24:54Z', info:NoneType=None,
    calc:bool=False
):
```

*Retrieves funding rate history from Hyperliquid for a given coin and time period.*

Args:
    coin (str, optional): Coin symbol (e.g. "ETH"). Defaults to "ETH".
    end_date (str, optional): End datetime in ISO 8601 format. Defaults to current UTC time.
    start_date (str, optional): Start datetime in ISO 8601 format. Defaults to 7 days before end_date.
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.

Returns:
    pandas.DataFrame: DataFrame containing the funding history with columns:
        - datetime: Timestamp for the funding rate (UTC)
        - funding_rate: The funding rate value
        - premium: The premium component
        - coin: Coin symbol
    Returns None if the API request fails or returns no data.

Notes:
    - All datetime values are in UTC timezone
    - Funding rates are typically updated every hour
    - Requires Hyperliquid Info client to be initialized

### Example

In [ ]:
#| eval: false
f_r = retrieve_hyperliquid_funding_history(info=info,calc=True)
print(f_r.tail())

                  datetime  funding_rate   premium coin  fund_calc
43 2025-12-04 20:00:00.021      0.000009 -0.000429  ETH   0.000009
44 2025-12-04 21:00:00.078      0.000010 -0.000422  ETH   0.000010
45 2025-12-04 22:00:00.034      0.000009 -0.000427  ETH   0.000009
46 2025-12-04 23:00:00.015      0.000007 -0.000442  ETH   0.000007
47 2025-12-05 00:00:00.002      0.000006 -0.000452  ETH   0.000006


## Unified function for easy data retrieval
### retrieve_hyperliquid_data function

In [0]:
#| echo: false
#| output: asis
show_doc(retrieve_hyperliquid_data)

---

### retrieve_hyperliquid_data

```python
def retrieve_hyperliquid_data(
    ticker:str='ETH', data_type:str='perp', start_date:NoneType=None, end_date:NoneType=None, lookback:int=2,
    interval:str='1h', base:str='USDC', round_to_hour:bool=False, info:NoneType=None
):
```

*Unified function to retrieve funding rates, perpetual prices, or spot prices from Hyperliquid.*

Args:
    ticker (str, optional): Coin symbol (e.g. "ETH", "BTC"). Defaults to "ETH".
    data_type (str, optional): Type of data to retrieve - "funding", "perp", or "spot". Defaults to "perp".
    start_date (str, optional): Start date as string. Can be:
        - ISO format: "2024-01-15T10:30:00Z" or "2024-01-15T10:30:00"
        - Date only: "2024-01-15"
        - If None, calculated from lookback. Defaults to None.
    end_date (str, optional): End date as string (same formats as start_date).
        - If None, uses current UTC time. Defaults to None.
    lookback (int, optional): Number of days to look back from end_date if start_date is None. 
        Defaults to 2.
    interval (str, optional): Candle interval for perp/spot data ("1m", "5m", "15m", "1h", "4h", "1d"). 
        Defaults to "1h". Not used for funding rates.
    base (str, optional): Base currency for spot prices (e.g. "USDC"). Defaults to "USDC".
        Not used for perp or funding rates.
    round_to_hour (bool, optional): If True, rounds start_date and end_date to nearest hour.
        Useful for funding rates which update hourly. Defaults to False.
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.

Returns:
    pandas.DataFrame: DataFrame containing the requested data with appropriate columns:
        - For "funding": datetime, funding_rate, premium, coin
        - For "perp": datetime, open, high, low, close, volume, coin
        - For "spot": datetime, open, high, low, close, volume, coin
    Returns None if the API request fails or returns no data.

Examples:
    # Get 7 days of funding rates for ETH, rounded to hour
    df = retrieve_hyperliquid_data("ETH", "funding", lookback=7, round_to_hour=True, info=info)

    # Get perp prices between specific dates with 4h interval
    df = retrieve_hyperliquid_data("BTC", "perp", 
                                  start_date="2024-01-01", 
                                  end_date="2024-01-15",
                                  interval="4h", info=info)

    # Get spot prices for last 30 days with 1h interval
    df = retrieve_hyperliquid_data("ETH", "spot", lookback=30, 
                                  interval="1h", base="USDC", info=info)

Notes:
    - All datetime values are in UTC timezone
    - Valid data_type values: "funding", "perp", "spot"
    - Funding rates are updated hourly, so round_to_hour=True is recommended
    - Requires Hyperliquid Info client to be initialized

### Example

In [ ]:
#| eval: false
## Example 1: Get funding rates for last 7 days, rounded to hour
funding_df = retrieve_hyperliquid_data(
    ticker="ETH",
    data_type="funding",
    lookback=7,
    round_to_hour=True,
    info=info
)
print(funding_df.tail())

               datetime  funding_rate   premium coin  fund_calc
163 2025-12-04 20:00:00      0.000009 -0.000429  ETH   0.000009
164 2025-12-04 21:00:00      0.000010 -0.000422  ETH   0.000010
165 2025-12-04 22:00:00      0.000009 -0.000427  ETH   0.000009
166 2025-12-04 23:00:00      0.000007 -0.000442  ETH   0.000007
167 2025-12-05 00:00:00      0.000006 -0.000452  ETH   0.000006


In [ ]:
#| eval: false
## Example 2: Get perpetual prices with specific dates and 4h interval
perp_df = retrieve_hyperliquid_data(
    ticker="BTC",
    data_type="perp",
    start_date="2024-01-01",
    end_date="2024-01-15",
    interval="4h",
    info=info
)
print(perp_df.tail())

              datetime     open     high      low    close     volume coin
80 2024-01-14 08:00:00  43034.0  43114.0  42786.0  42862.0  158.38267  BTC
81 2024-01-14 12:00:00  42858.0  43019.0  42743.0  42936.0  261.86230  BTC
82 2024-01-14 16:00:00  42936.0  43035.0  42681.0  42698.0  304.01708  BTC
83 2024-01-14 20:00:00  42699.0  42772.0  41762.0  41786.0  721.09226  BTC
84 2024-01-15 00:00:00  41779.0  42674.0  41737.0  42612.0  281.41095  BTC


In [ ]:
#| eval: false
## Example 3: Get spot prices for last 30 days
spot_df = retrieve_hyperliquid_data(
    ticker="ETH",
    data_type="spot",
    lookback=30,
    interval="1h",
    base="USDC",
    info=info
)
print(spot_df.tail())

               datetime    open    high     low   close    volume coin
716 2025-12-04 20:00:00  3120.6  3145.6  3106.6  3143.1  202.4952  ETH
717 2025-12-04 21:00:00  3141.0  3154.2  3120.0  3123.9  339.5227  ETH
718 2025-12-04 22:00:00  3125.1  3148.6  3124.4  3144.5   44.5737  ETH
719 2025-12-04 23:00:00  3141.9  3144.7  3130.9  3133.9   54.7757  ETH
720 2025-12-05 00:00:00  3133.9  3156.3  3133.4  3147.9   35.9388  ETH


In [ ]:
#| eval: false
## Example 4: Using datetime strings with time information
data_df = retrieve_hyperliquid_data(
    ticker="ETH",
    data_type="perp",
    lookback=2,
    interval="1h",
    info=info
)
print(data_df.tail())

              datetime    open    high     low   close      volume coin
44 2025-12-04 20:00:00  3117.9  3143.8  3104.4  3140.5  15553.2974  ETH
45 2025-12-04 21:00:00  3140.2  3150.0  3117.2  3121.8  13448.5987  ETH
46 2025-12-04 22:00:00  3122.1  3147.2  3121.5  3142.4  11888.6633  ETH
47 2025-12-04 23:00:00  3142.6  3142.6  3127.9  3131.7   4731.0506  ETH
48 2025-12-05 00:00:00  3131.8  3156.0  3131.4  3145.2  10706.1671  ETH


In [ ]:
#| eval: false
## Example 5: Get funding rates with date-only strings
funding_df = retrieve_hyperliquid_data(
    ticker="BTC",
    data_type="funding",
    start_date="2025-08-29",
    end_date="2025-09-01",
    round_to_hour=True,
    info=info
)
print(funding_df)

              datetime  funding_rate   premium coin  fund_calc
0  2025-08-29 00:00:00      0.000013  0.000251  BTC   0.000013
1  2025-08-29 01:00:00      0.000013  0.000234  BTC   0.000013
2  2025-08-29 02:00:00      0.000013  0.000124  BTC   0.000013
3  2025-08-29 03:00:00      0.000013  0.000068  BTC   0.000013
4  2025-08-29 04:00:00      0.000013  0.000086  BTC   0.000013
..                 ...           ...       ...  ...        ...
67 2025-08-31 19:00:00      0.000013 -0.000035  BTC   0.000013
68 2025-08-31 20:00:00      0.000013 -0.000082  BTC   0.000013
69 2025-08-31 21:00:00      0.000013 -0.000028  BTC   0.000013
70 2025-08-31 22:00:00      0.000013 -0.000059  BTC   0.000013
71 2025-08-31 23:00:00      0.000013 -0.000012  BTC   0.000013

[72 rows x 5 columns]


In [ ]:
#| eval: false
data_df = retrieve_hyperliquid_data(
    ticker="ETH",
    data_type="perp",
    start_date="2025-06-15T10:30:00",
    end_date="2025-06-20T15:45:00",
    interval="1h",
    info=info
)
data_df.tail()

,datetime,open,high,low,close,volume,coin
121,2025-06-20 11:00:00,2551.4,2555.0,2540.1,2548.2,27254.2428,ETH
122,2025-06-20 12:00:00,2548.1,2557.0,2546.1,2550.7,15547.1122,ETH
123,2025-06-20 13:00:00,2550.6,2558.4,2527.5,2534.9,32702.0042,ETH
124,2025-06-20 14:00:00,2535.0,2539.6,2488.1,2488.5,49493.5297,ETH
125,2025-06-20 15:00:00,2488.4,2506.8,2486.1,2492.1,42340.1040,ETH


## Order book data
### retrieve_hyperliquid_l2_snapshot

The will return the snapshot of the order book for the specified ticker in the moment you call the function. It does not provide historical data. 

This is good only for perpetual markets.

In [ ]:
def unix_to_datetime(t):
    return datetime.fromtimestamp(t/1000).strftime("%Y-%m-%d %H:%M:%S.%f")
#The funding rates are reset every hour.
#t = datetime.datetime.fromtimestamp(i['time']/1000).strftime("%Y-%m-%d %H:%M:%S.%f")
#print(t)

In [0]:
#| echo: false
#| output: asis
show_doc(retrieve_hyperliquid_l2_snapshot)

---

### retrieve_hyperliquid_l2_snapshot

```python
def retrieve_hyperliquid_l2_snapshot(
    coin:str='ETH', info:NoneType=None
):
```

*Retrieves current L2 order book snapshot from Hyperliquid for a given coin.*

Args:
    coin (str, optional): Coin symbol (e.g. "ETH", "BTC"). Defaults to "ETH".
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.

Returns:
    pandas.DataFrame: DataFrame containing the order book snapshot with columns:
        - datetime: Timestamp of the snapshot (UTC)
        - side: Order side ("bid" or "ask")
        - price: Price level
        - size: Total size at this price level
        - num_orders: Number of orders at this price level
    Returns None if the API request fails or returns no data.

Notes:
    - This is a snapshot at the moment the function is called
    - All datetime values are in UTC timezone
    - Bids are sorted from highest to lowest price
    - Asks are sorted from lowest to highest price
    - Requires Hyperliquid Info client to be initialized

### Example

In [ ]:
#| eval: false
# Example usage: Get L2 order book snapshot for ETH
l2_snapshot = retrieve_hyperliquid_l2_snapshot(coin="ETH", info=info)
print(l2_snapshot.head(5))
print(l2_snapshot.tail(5))

# Check the structure
print(f"\nTotal levels: {len(l2_snapshot)}")
print(f"Bids: {len(l2_snapshot[l2_snapshot['side'] == 'bid'])}")
print(f"Asks: {len(l2_snapshot[l2_snapshot['side'] == 'ask'])}")
print(f"Snapshot time: {l2_snapshot['datetime'].iloc[0]}")

                 datetime side   price     size  num_orders
0 2025-12-05 00:29:21.230  bid  3145.1  27.1027          10
1 2025-12-05 00:29:21.230  bid  3145.0  36.0446           5
2 2025-12-05 00:29:21.230  bid  3144.9  35.2263           4
3 2025-12-05 00:29:21.230  bid  3144.8  73.4864           7
4 2025-12-05 00:29:21.230  bid  3144.7  33.8558           7
                  datetime side   price      size  num_orders
35 2025-12-05 00:29:21.230  ask  3146.8  227.1072          10
36 2025-12-05 00:29:21.230  ask  3146.9   65.8934           5
37 2025-12-05 00:29:21.230  ask  3147.0  409.3202          42
38 2025-12-05 00:29:21.230  ask  3147.1  360.0039          11
39 2025-12-05 00:29:21.230  ask  3147.2   30.0297           3

Total levels: 40
Bids: 20
Asks: 20
Snapshot time: 2025-12-05 00:29:21.230000


## Mid prices
### hyperliquid_mids

In [0]:
#| echo: false
#| output: asis
show_doc(hyperliquid_mids)

---

### hyperliquid_mids

```python
def hyperliquid_mids(
    coin:NoneType=None, info:NoneType=None, typecast_to_float:bool=True
):
```

*Retrieves current mid prices from Hyperliquid for specified coins or all available coins.*

Args:
    coins (list, optional): List of coin symbols (e.g. ["ETH", "BTC"]). 
        If None, retrieves mids for all available coins. Defaults to None.
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.

Returns:
    dictionary: Dictionary containing mid prices with keys as coin symbols and values as mid prices
    Returns None if the API request fails or returns no data.

Examples:
    # Get mid prices for specific coins
    dic = hyperliquid_mids(info=info)

Notes:
    - This is a snapshot at the moment the function is called
    - Mid price is calculated as (best_bid + best_ask) / 2
    - Requires Hyperliquid Info client to be initialized

### Example

In [ ]:
#| eval: false
hyperliquid_mids(coin="ETH",info=info)

3145.15

In [ ]:
#| eval: false
hyperliquid_mids(info=info)

{'0G': 1.1473,
 '2Z': 0.130635,
 '@1': 22.784,
 '@10': 3.05e-05,
 '@100': 0.004374,
 '@101': 0.21451,
 '@102': 0.018967,
 '@103': 3.534e-05,
 '@104': 0.064286,
 '@105': 0.379665,
 '@106': 0.009938,
 '@107': 33.4665,
 '@108': 0.012217,
 '@109': 0.00028937,
 '@11': 0.000584,
 '@110': 0.01078,
 '@111': 0.044361,
 '@112': 0.0006262,
 '@113': 0.0001106,
 '@114': 0.0002028,
 '@115': 0.053924,
 '@116': 2.334e-05,
 '@117': 0.0011372,
 '@118': 0.014123,
 '@119': 0.014502,
 '@12': 4.55e-05,
 '@120': 0.016756,
 '@121': 0.006519,
 '@122': 0.009554,
 '@123': 0.084191,
 '@124': 0.006077,
 '@125': 0.004406,
 '@126': 0.085535,
 '@127': 0.10031,
 '@128': 0.001006,
 '@129': 0.010247,
 '@13': 0.00091893,
 '@130': 2.2e-05,
 '@131': 0.07475,
 '@132': 0.002749,
 '@133': 0.000261,
 '@134': 0.000944,
 '@135': 1.4e-05,
 '@136': 0.015315,
 '@137': 1.7e-05,
 '@138': 0.0009317,
 '@139': 0.042716,
 '@14': 7.482e-05,
 '@140': 0.00026,
 '@141': 0.031348,
 '@142': 92361.5,
 '@143': 0.005822,
 '@144': 8.3e-05,
 '@145'

## Data management

Similar to coinbase.py, this script saves the data in a specified format. These functions handles duplicates and sorts by date. For hourly data, it aligns to the hour.

### HyperliquidDataManager base class

In [0]:
#| echo: false
#| output: asis
show_doc(save_hyperliquid_file)

---

### save_hyperliquid_file

```python
def save_hyperliquid_file(
    df, folder_path, file_name, type:str='parquet'
):
```

*Save a pandas DataFrame to a file in either CSV or Parquet format.*

Args:
    df (pandas.DataFrame): The DataFrame to save
    folder_path (str): Directory path where the file will be saved
    file_name (str): Name of the file without extension
    type (str, optional): File format - either "csv" or "parquet". Defaults to "parquet"

The function saves the DataFrame to the specified path, handling the file extension automatically.
For CSV files, the index is not saved. For Parquet files, default Parquet settings are used.
Creates the folder if it doesn't exist.

In [0]:
#| echo: false
#| output: asis
show_doc(HyperliquidDataManager)

---

### HyperliquidDataManager

```python
def HyperliquidDataManager(
    ticker:NoneType=None, data_dir:str='../data/hyperliquid', interval:str='1h', file_type:str='parquet',
    update:bool=False, save:bool=False, refresh_hours:int=24, info:NoneType=None, verbose:bool=True,
    data_type:NoneType=None
):
```

*Base class for managing Hyperliquid data files.*

Handles reading, updating, and saving data for different data types (perp, spot, funding).
Data is stored in organized folders by data type, with files named by token and interval.

Args:
    ticker (str or list, optional): Token symbol(s) to manage. If None, uses all available tokens.
    data_dir (str, optional): Base directory for data storage. Defaults to "../data/hyperliquid"
    interval (str, optional): Time interval for data ("1m", "5m", "15m", "1h", "4h", "1d"). 
        Defaults to "1h". Not used for funding data.
    file_type (str, optional): File format - "parquet" or "csv". Defaults to "parquet"
    update (bool, optional): If True, checks for and downloads new data. Defaults to False
    save (bool, optional): If True, saves updated data back to file. Defaults to False
    refresh_hours (int, optional): Hours of data to refresh when updating. Defaults to 24
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.
    verbose (bool, optional): If True, prints progress messages. Defaults to True

Attributes:
    data_type (str): Type of data managed by this instance ("perp", "spot", "funding")
    data (dict): Dictionary mapping tickers to their DataFrames

### HyperliquidPerpManager derived class

In [0]:
#| echo: false
#| output: asis
show_doc(HyperliquidPerpManager)

---

### HyperliquidPerpManager

```python
def HyperliquidPerpManager(
    ticker:NoneType=None, data_dir:str='../data/hyperliquid', interval:str='1h', file_type:str='parquet',
    update:bool=False, save:bool=False, refresh_hours:int=24, info:NoneType=None, verbose:bool=True
):
```

*Manager for Hyperliquid perpetual futures data.*

Handles reading, updating, and saving perpetual price data (OHLCV).
Data is stored in the 'perp' subfolder with files named as {ticker}_{interval}.{file_type}

Args:
    ticker (str or list, optional): Token symbol(s) to manage. If None, uses all available perp tokens.
    data_dir (str, optional): Base directory for data storage. Defaults to "../data/hyperliquid"
    interval (str, optional): Time interval for data ("1m", "5m", "15m", "1h", "4h", "1d"). 
        Defaults to "1h".
    file_type (str, optional): File format - "parquet" or "csv". Defaults to "parquet"
    update (bool, optional): If True, checks for and downloads new data. Defaults to False
    save (bool, optional): If True, saves updated data back to file. Defaults to False
    refresh_hours (int, optional): Hours of data to refresh when updating. Defaults to 24
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.
    verbose (bool, optional): If True, prints progress messages. Defaults to True

Examples:
    # Load existing perp data for ETH
    manager = HyperliquidPerpManager(ticker="ETH", interval="1h", info=info)
    eth_data = manager.data["ETH"]

    # Update and save data for multiple tokens
    manager = HyperliquidPerpManager(
        ticker=["ETH", "BTC", "SOL"],
        interval="4h",
        update=True,
        save=True,
        refresh_hours=48,
        info=info
    )

    # Load all available perp tokens
    manager = HyperliquidPerpManager(update=True, save=True, info=info)

#### Example

In [ ]:
#| eval: false
# Load and update ETH perpetual data with 1h interval
manager = HyperliquidPerpManager(
    ticker="ETH",
    interval="1h",
    refresh_hours = 24,
    update=True,
    save=True,
    info=info
)
eth_data = manager.get_data("ETH")
print(eth_data)

Processing 1 perpetual tokens...
Loaded 5888 rows for ETH from ../data/hyperliquid/perp/ETH_1h.parquet
Updating ETH from 2025-11-18T22:00:00Z
Retrieved 387 new rows for ETH
Added 362 new rows for ETH
Saved 6250 rows for ETH
                datetime    open    high     low   close      volume coin
0    2025-03-19 15:00:00  2030.7  2056.7  2026.0  2048.7  30206.2723  ETH
1    2025-03-19 16:00:00  2048.7  2049.2  2035.3  2047.4  22198.4803  ETH
2    2025-03-19 17:00:00  2047.4  2048.9  2014.3  2026.1  35909.8007  ETH
3    2025-03-19 18:00:00  2026.1  2059.9  1998.0  2045.0  84148.3569  ETH
4    2025-03-19 19:00:00  2045.1  2052.1  2020.1  2029.7  38921.5619  ETH
...                  ...     ...     ...     ...     ...         ...  ...
6245 2025-12-04 20:00:00  3117.9  3143.8  3104.4  3140.5  15553.2974  ETH
6246 2025-12-04 21:00:00  3140.2  3150.0  3117.2  3121.8  13448.5987  ETH
6247 2025-12-04 22:00:00  3122.1  3147.2  3121.5  3142.4  11888.6633  ETH
6248 2025-12-04 23:00:00  3142.6  31

In [ ]:
#|eval: false
manager = HyperliquidPerpManager(
    update=False,
    save=False,
    verbose=False
)
df = manager.get_data()
print(df)

                      datetime     open     high      low    close    volume  \
0    2025-10-30 19:00:00+00:00  0.24613  0.24678  0.24364  0.24403   33184.0   
1    2025-10-30 20:00:00+00:00  0.24454  0.24729  0.24406  0.24718   13245.0   
2    2025-10-30 21:00:00+00:00  0.24716  0.24958  0.24716  0.24886   11172.0   
3    2025-10-30 22:00:00+00:00  0.24922  0.25149  0.24905  0.25148    5984.0   
4    2025-10-30 23:00:00+00:00  0.25177  0.25497  0.25089  0.25467    5767.0   
...                        ...      ...      ...      ...      ...       ...   
5885 2025-11-19 18:00:00+00:00  0.33057  0.34508  0.32577  0.33751  952658.5   
5886 2025-11-19 19:00:00+00:00  0.33635  0.35653  0.33635  0.35555  120968.8   
5887 2025-11-19 20:00:00+00:00  0.35623  0.37132  0.35614  0.36231  289579.7   
5888 2025-11-19 21:00:00+00:00  0.36306  0.38125  0.36249  0.37841  361439.4   
5889 2025-11-19 22:00:00+00:00  0.37841  0.38556  0.37205  0.38395  131409.8   

       coin  
0     LAYER  
1     LAYER

### HyperliquidSpotManager derived class

In [0]:
#| echo: false
#| output: asis
show_doc(HyperliquidSpotManager)

---

### HyperliquidSpotManager

```python
def HyperliquidSpotManager(
    ticker:NoneType=None, base:str='USDC', data_dir:str='../data/hyperliquid', interval:str='1h',
    file_type:str='parquet', update:bool=False, save:bool=False, refresh_hours:int=24, info:NoneType=None,
    verbose:bool=True
):
```

*Manager for Hyperliquid spot market data.*

Handles reading, updating, and saving spot price data (OHLCV).
Data is stored in the 'spot' subfolder with files named as {ticker}_{base}_{interval}.{file_type}

Args:
    ticker (str or list, optional): Token symbol(s) to manage. If None, uses all available spot tokens.
    base (str, optional): Base currency for spot pairs. Defaults to "USDC".
    data_dir (str, optional): Base directory for data storage. Defaults to "../data/hyperliquid"
    interval (str, optional): Time interval for data ("1m", "5m", "15m", "1h", "4h", "1d"). 
        Defaults to "1h".
    file_type (str, optional): File format - "parquet" or "csv". Defaults to "parquet"
    update (bool, optional): If True, checks for and downloads new data. Defaults to False
    save (bool, optional): If True, saves updated data back to file. Defaults to False
    refresh_hours (int, optional): Hours of data to refresh when updating. Defaults to 24
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.
    verbose (bool, optional): If True, prints progress messages. Defaults to True

Examples:
    # Load existing spot data for ETH/USDC
    manager = HyperliquidSpotManager(ticker="ETH", base="USDC", interval="1h", info=info)
    eth_data = manager.data["ETH"]

    # Update and save data for multiple tokens
    manager = HyperliquidSpotManager(
        ticker=["ETH", "BTC", "SOL"],
        base="USDC",
        interval="4h",
        update=True,
        save=True,
        refresh_hours=48,
        info=info
    )

    # Load all available spot tokens
    manager = HyperliquidSpotManager(base="USDC", update=True, save=True, info=info)

#### Example

In [ ]:
#| eval: false
# Example 1: Load existing spot data for a single token
print("Example 1: Load existing ETH/USDC spot data")
manager1 = HyperliquidSpotManager(
    ticker="ETH",
    base="USDC",
    interval="1h",
    update=True,
    save=True,
    info=info,
    verbose=True
)
if "ETH" in manager1.data:
    print(f"Loaded {len(manager1.data['ETH'])} records for ETH")
    print(manager1.data["ETH"])

Example 1: Load existing ETH/USDC spot data
Loaded 5487 rows for ETH from ../data/hyperliquid/spot/ETH_USDC_1h.parquet
Updating ETH/USDC from 2025-11-18T22:00:00Z
Retrieved 387 new rows for ETH/USDC
Added 362 new rows for ETH/USDC
Saved 5849 rows for ETH/USDC
Loaded 5849 records for ETH
                datetime    open    high     low   close    volume coin
0    2025-04-05 08:00:00  1812.3  1813.9  1809.7  1812.7    1.7000  ETH
1    2025-04-05 09:00:00  1814.0  1821.0  1814.0  1819.9   19.6808  ETH
2    2025-04-05 10:00:00  1820.2  1820.2  1816.2  1817.4   12.5505  ETH
3    2025-04-05 11:00:00  1818.4  1818.5  1806.6  1806.8   63.9187  ETH
4    2025-04-05 12:00:00  1806.8  1806.9  1793.0  1796.5   45.5064  ETH
...                  ...     ...     ...     ...     ...       ...  ...
5844 2025-12-04 20:00:00  3120.6  3145.6  3106.6  3143.1  202.4952  ETH
5845 2025-12-04 21:00:00  3141.0  3154.2  3120.0  3123.9  339.5227  ETH
5846 2025-12-04 22:00:00  3125.1  3148.6  3124.4  3144.5   44.57

In [ ]:
#| eval: false
# Example 2: Update and save data for multiple tokens
print("\nExample 2: Update and save multiple tokens")
manager2 = HyperliquidSpotManager(
    ticker=["ETH", "BTC", "SOL"],
    base="USDC",
    interval="1h",
    update=True,
    save=True,
    refresh_hours=24,
    info=info,
    verbose=True
)

# Access the data
for ticker in ["ETH", "BTC", "SOL"]:
    if ticker in manager2.data:
        print(f"\n{ticker} data shape: {manager2.data[ticker].shape}")
        print(f"Latest {ticker} price: ${manager2.data[ticker]['close'].iloc[-1]:.2f}")


Example 2: Update and save multiple tokens
Loaded 5849 rows for ETH from ../data/hyperliquid/spot/ETH_USDC_1h.parquet
Updating ETH/USDC from 2025-12-04T00:00:00Z
Retrieved 25 new rows for ETH/USDC
Added 0 new rows for ETH/USDC
Saved 5849 rows for ETH/USDC
Loaded 5487 rows for BTC from ../data/hyperliquid/spot/BTC_USDC_1h.parquet
Updating BTC/USDC from 2025-11-18T22:00:00Z
Retrieved 387 new rows for BTC/USDC
Added 362 new rows for BTC/USDC
Saved 5849 rows for BTC/USDC
Loaded 4647 rows for SOL from ../data/hyperliquid/spot/SOL_USDC_1h.parquet
Updating SOL/USDC from 2025-11-18T22:00:00Z
Retrieved 387 new rows for SOL/USDC
Added 362 new rows for SOL/USDC
Saved 5009 rows for SOL/USDC

ETH data shape: (5849, 7)
Latest ETH price: $3147.60

BTC data shape: (5849, 7)
Latest BTC price: $92361.00

SOL data shape: (5009, 7)
Latest SOL price: $139.50


In [ ]:
#| eval: false
# Example 3: Load all available spot tokens with USDC base
print("\nExample 3: Load all available spot tokens")
manager3 = HyperliquidSpotManager(
    base="USDC",
    interval="4h",
    update=True,
    save=True,
    info=info,
    verbose=True
)

print(f"\nLoaded {len(manager3.data)} spot tokens")
print(f"Available tokens: {list(manager3.data.keys())}")


Example 3: Load all available spot tokens
Found 7 spot tokens for USDC
Loaded 105 rows for BTC from ../data/hyperliquid/spot/BTC_USDC_4h.parquet
Updating BTC/USDC from 2025-11-14T20:00:00Z
Retrieved 122 new rows for BTC/USDC
Added 115 new rows for BTC/USDC
Saved 220 rows for BTC/USDC
Loaded 105 rows for ETH from ../data/hyperliquid/spot/ETH_USDC_4h.parquet
Updating ETH/USDC from 2025-11-14T20:00:00Z
Retrieved 122 new rows for ETH/USDC
Added 115 new rows for ETH/USDC
Saved 220 rows for ETH/USDC
Loaded 105 rows for SOL from ../data/hyperliquid/spot/SOL_USDC_4h.parquet
Updating SOL/USDC from 2025-11-14T20:00:00Z
Retrieved 122 new rows for SOL/USDC
Added 115 new rows for SOL/USDC
Saved 220 rows for SOL/USDC
Loaded 105 rows for HYPE from ../data/hyperliquid/spot/HYPE_USDC_4h.parquet
Updating HYPE/USDC from 2025-11-14T20:00:00Z
Retrieved 122 new rows for HYPE/USDC
Added 115 new rows for HYPE/USDC
Saved 220 rows for HYPE/USDC
Loaded 43 rows for TRUMP from ../data/hyperliquid/spot/TRUMP_USDC_

In [ ]:
#| eval: false
# Example 5: Analyze spot data
print("\nExample 5: Analyze spot data")
manager5 = HyperliquidSpotManager(
    ticker=["ETH", "BTC"],
    base="USDC",
    interval="1h",
    update=True,
    info=info,
    verbose=True
)

# Calculate some basic statistics
for ticker in ["ETH", "BTC"]:
    if ticker in manager5.data:
        df = manager5.data[ticker]
        
        # Calculate 24h change
        if len(df) >= 24:
            price_24h_ago = df['close'].iloc[-24]
            current_price = df['close'].iloc[-1]
            change_24h = ((current_price - price_24h_ago) / price_24h_ago) * 100
            
            print(f"\n{ticker}/USDC:")
            print(f"  Current Price: ${current_price:.2f}")
            print(f"  24h Change: {change_24h:+.2f}%")
            print(f"  24h High: ${df['high'].iloc[-24:].max():.2f}")
            print(f"  24h Low: ${df['low'].iloc[-24:].min():.2f}")
            print(f"  24h Volume: {df['volume'].iloc[-24:].sum():.2f}")


Example 5: Analyze spot data
Loaded 5849 rows for ETH from ../data/hyperliquid/spot/ETH_USDC_1h.parquet
Updating ETH/USDC from 2025-12-04T00:00:00Z
Retrieved 25 new rows for ETH/USDC
Added 0 new rows for ETH/USDC
Loaded 5849 rows for BTC from ../data/hyperliquid/spot/BTC_USDC_1h.parquet
Updating BTC/USDC from 2025-12-04T00:00:00Z
Retrieved 25 new rows for BTC/USDC
Added 0 new rows for BTC/USDC

ETH/USDC:
  Current Price: $3147.60
  24h Change: -1.75%
  24h High: $3239.80
  24h Low: $3069.80
  24h Volume: 5202.40

BTC/USDC:
  Current Price: $92361.00
  24h Change: -0.91%
  24h High: $94081.00
  24h Low: $90852.00
  24h Volume: 587.53


In [ ]:
#|eval: false
mananger6 = HyperliquidSpotManager(
    base="USDC",
    interval="1h",
    update=False,
    save=False,
    verbose=True
)
mananger6.get_data()

Found 7 spot tokens for USDC
Loaded 5849 rows for BTC from ../data/hyperliquid/spot/BTC_USDC_1h.parquet
Loaded 5849 rows for ETH from ../data/hyperliquid/spot/ETH_USDC_1h.parquet
Loaded 5009 rows for SOL from ../data/hyperliquid/spot/SOL_USDC_1h.parquet
Loaded 484 rows for HYPE from ../data/hyperliquid/spot/HYPE_USDC_1h.parquet
Loaded 36 rows for TRUMP from ../data/hyperliquid/spot/TRUMP_USDC_1h.parquet
Loaded 25 rows for BERA from ../data/hyperliquid/spot/BERA_USDC_1h.parquet
Loaded 480 rows for PUMP from ../data/hyperliquid/spot/PUMP_USDC_1h.parquet


,datetime,open,high,low,close,volume,coin
0,2025-04-05 08:00:00+00:00,83511.000000,83623.000000,83361.000000,83482.000000,2.416920e+00,BTC
1,2025-04-05 09:00:00+00:00,83486.000000,83854.000000,83486.000000,83848.000000,6.817760e+00,BTC
2,2025-04-05 10:00:00+00:00,83855.000000,83901.000000,83600.000000,83642.000000,6.209230e+00,BTC
3,2025-04-05 11:00:00+00:00,83642.000000,83688.000000,83402.000000,83430.000000,9.042830e+00,BTC
4,2025-04-05 12:00:00+00:00,83441.000000,83517.000000,83112.000000,83243.000000,6.552230e+00,BTC
...,...,...,...,...,...,...,...
475,2025-11-19 13:00:00+00:00,0.000148,0.000148,0.000148,0.000148,0.000000e+00,PUMP
476,2025-11-19 14:00:00+00:00,0.000148,0.000151,0.000148,0.000151,1.585506e+06,PUMP
477,2025-11-19 15:00:00+00:00,0.000151,0.000151,0.000151,0.000151,0.000000e+00,PUMP
478,2025-11-19 16:00:00+00:00,0.000151,0.000151,0.000151,0.000151,0.000000e+00,PUMP


### HyperliquidFundingManager derived class

In [0]:
#| echo: false
#| output: asis
show_doc(HyperliquidFundingManager)

---

### HyperliquidFundingManager

```python
def HyperliquidFundingManager(
    ticker:NoneType=None, data_dir:str='../data/hyperliquid', file_type:str='parquet', update:bool=False,
    save:bool=False, refresh_hours:int=24, round_to_hour:bool=True, info:NoneType=None, verbose:bool=True
):
```

*Manager for Hyperliquid funding rate data.*

Handles reading, updating, and saving funding rate data.
Data is stored in the 'funding' subfolder with files named as {ticker}.{file_type}

Args:
    ticker (str or list, optional): Token symbol(s) to manage. If None, uses all available tokens.
    data_dir (str, optional): Base directory for data storage. Defaults to "../data/hyperliquid"
    file_type (str, optional): File format - "parquet" or "csv". Defaults to "parquet"
    update (bool, optional): If True, checks for and downloads new data. Defaults to False
    save (bool, optional): If True, saves updated data back to file. Defaults to False
    refresh_hours (int, optional): Hours of data to refresh when updating. Defaults to 24
    round_to_hour (bool, optional): If True, rounds datetime to nearest hour. Defaults to True
    info (Info, optional): Hyperliquid Info client. If None, creates a new one.
    verbose (bool, optional): If True, prints progress messages. Defaults to True

Examples:
    # Load existing funding data for ETH
    manager = HyperliquidFundingManager(ticker="ETH", info=info)
    eth_data = manager.data["ETH"]

    # Update and save data for multiple tokens
    manager = HyperliquidFundingManager(
        ticker=["ETH", "BTC", "SOL"],
        update=True,
        save=True,
        refresh_hours=48,
        info=info
    )

    # Load all available tokens
    manager = HyperliquidFundingManager(update=True, save=True, info=info)

#### Example

In [ ]:
#|eval: false
manager = HyperliquidFundingManager(ticker="ETH")
eth_data = manager.data["ETH"]
print(eth_data)

Loaded 738 rows for ETH from ../data/hyperliquid/funding/ETH.parquet
                     datetime  funding_rate   premium coin  fund_calc
0   2025-10-03 21:00:00+00:00      0.000013  0.000185  ETH   0.000013
1   2025-10-03 22:00:00+00:00      0.000013  0.000111  ETH   0.000013
2   2025-10-03 23:00:00+00:00      0.000013  0.000141  ETH   0.000013
3   2025-10-04 00:00:00+00:00      0.000013  0.000113  ETH   0.000013
4   2025-10-04 01:00:00+00:00      0.000013  0.000081  ETH   0.000013
..                        ...           ...       ...  ...        ...
733 2025-11-19 07:00:00+00:00      0.000013  0.000175  ETH   0.000013
734 2025-11-19 08:00:00+00:00      0.000013  0.000281  ETH   0.000013
735 2025-11-19 09:00:00+00:00      0.000013  0.000277  ETH   0.000013
736 2025-11-19 10:00:00+00:00      0.000013  0.000256  ETH   0.000013
737 2025-11-19 11:00:00+00:00      0.000013  0.000295  ETH   0.000013

[738 rows x 5 columns]


In [ ]:
#|eval: false
manager = HyperliquidFundingManager(update=False, save=False,verbose=False)
manager.get_data()

,datetime,funding_rate,premium,coin,fund_calc
0,2025-10-03 21:00:00+00:00,1.250000e-05,-0.000022,LAYER,1.250000e-05
1,2025-10-03 22:00:00+00:00,1.250000e-05,-0.000036,LAYER,1.250000e-05
2,2025-10-03 23:00:00+00:00,1.250000e-05,-0.000078,LAYER,1.250000e-05
3,2025-10-04 00:00:00+00:00,1.250000e-05,-0.000011,LAYER,1.250000e-05
4,2025-10-04 01:00:00+00:00,1.250000e-05,-0.000026,LAYER,1.250000e-05
...,...,...,...,...,...
1114,2025-11-19 07:00:00+00:00,-1.590130e-05,-0.000627,GRASS,-1.590130e-05
1115,2025-11-19 08:00:00+00:00,-4.340640e-05,-0.000847,GRASS,-4.340640e-05
1116,2025-11-19 09:00:00+00:00,-2.114020e-05,-0.000669,GRASS,-2.114020e-05
1117,2025-11-19 10:00:00+00:00,-1.030000e-08,-0.000500,GRASS,-1.030000e-08


## More examples

Load package:

In [ ]:
#|eval: false
from token_data.hyperliquid import *

Load all available funding rates already stored in the data directory. It returns a pandas DataFrame with columns 'datetime' in UTC and 'funding_rate' as the funding rate used within the Hyperliquid system to find the USDC paid or received for a given perpetual position.

The other coluns are "premium" and "fund_calc". The "premium" column represents the premium used in the funding rate calculation and the "fund_calc" is currently very experimental and tries to replicated the funding rate calculation in the Hyperliquid system. Use "funding_rate" column in simulations or P&L calculations.

In [ ]:
#|eval: false
manager = HyperliquidFundingManager(update=False, save=False, verbose=False)
manager.get_data()

,datetime,funding_rate,premium,coin,fund_calc
0,2025-10-03 21:00:00+00:00,1.250000e-05,-0.000022,LAYER,1.250000e-05
1,2025-10-03 22:00:00+00:00,1.250000e-05,-0.000036,LAYER,1.250000e-05
2,2025-10-03 23:00:00+00:00,1.250000e-05,-0.000078,LAYER,1.250000e-05
3,2025-10-04 00:00:00+00:00,1.250000e-05,-0.000011,LAYER,1.250000e-05
4,2025-10-04 01:00:00+00:00,1.250000e-05,-0.000026,LAYER,1.250000e-05
...,...,...,...,...,...
1114,2025-11-19 07:00:00+00:00,-1.590130e-05,-0.000627,GRASS,-1.590130e-05
1115,2025-11-19 08:00:00+00:00,-4.340640e-05,-0.000847,GRASS,-4.340640e-05
1116,2025-11-19 09:00:00+00:00,-2.114020e-05,-0.000669,GRASS,-2.114020e-05
1117,2025-11-19 10:00:00+00:00,-1.030000e-08,-0.000500,GRASS,-1.030000e-08


Load all available spot prices already stored in the data directory. It returns a pandas DataFrame with columns 'datetime' in UTC and 'open', 'high', 'low', 'close', 'volume' and 'coin'. Important to note that the list of spots is limited and was hard coded. In the future this can be changed, therefore, the list is limited to some of the most important tokens.

In [ ]:
#|eval: false
manager = HyperliquidSpotManager(update=False, save=False, verbose=False)
manager.get_data()

,datetime,open,high,low,close,volume,coin
0,2025-04-05 08:00:00+00:00,83511.000000,83623.000000,83361.000000,83482.000000,2.416920e+00,BTC
1,2025-04-05 09:00:00+00:00,83486.000000,83854.000000,83486.000000,83848.000000,6.817760e+00,BTC
2,2025-04-05 10:00:00+00:00,83855.000000,83901.000000,83600.000000,83642.000000,6.209230e+00,BTC
3,2025-04-05 11:00:00+00:00,83642.000000,83688.000000,83402.000000,83430.000000,9.042830e+00,BTC
4,2025-04-05 12:00:00+00:00,83441.000000,83517.000000,83112.000000,83243.000000,6.552230e+00,BTC
...,...,...,...,...,...,...,...
475,2025-11-19 13:00:00+00:00,0.000148,0.000148,0.000148,0.000148,0.000000e+00,PUMP
476,2025-11-19 14:00:00+00:00,0.000148,0.000151,0.000148,0.000151,1.585506e+06,PUMP
477,2025-11-19 15:00:00+00:00,0.000151,0.000151,0.000151,0.000151,0.000000e+00,PUMP
478,2025-11-19 16:00:00+00:00,0.000151,0.000151,0.000151,0.000151,0.000000e+00,PUMP


Next example loads perpetual prices already stored in the data directory on an hourly frequency (default but can be changed). It returns a pandas DataFrame with columns 'datetime' in UTC and 'open', 'high', 'low', 'close', 'volume' and 'coin'.

In [ ]:
#|eval: false
manager = HyperliquidPerpManager(update=False, save=False,verbose=False)
manager.get_data()

,datetime,open,high,low,close,volume,coin
0,2025-10-30 19:00:00+00:00,0.24613,0.24678,0.24364,0.24403,33184.0,LAYER
1,2025-10-30 20:00:00+00:00,0.24454,0.24729,0.24406,0.24718,13245.0,LAYER
2,2025-10-30 21:00:00+00:00,0.24716,0.24958,0.24716,0.24886,11172.0,LAYER
3,2025-10-30 22:00:00+00:00,0.24922,0.25149,0.24905,0.25148,5984.0,LAYER
4,2025-10-30 23:00:00+00:00,0.25177,0.25497,0.25089,0.25467,5767.0,LAYER
...,...,...,...,...,...,...,...
5885,2025-11-19 18:00:00+00:00,0.33057,0.34508,0.32577,0.33751,952658.5,GRASS
5886,2025-11-19 19:00:00+00:00,0.33635,0.35653,0.33635,0.35555,120968.8,GRASS
5887,2025-11-19 20:00:00+00:00,0.35623,0.37132,0.35614,0.36231,289579.7,GRASS
5888,2025-11-19 21:00:00+00:00,0.36306,0.38125,0.36249,0.37841,361439.4,GRASS


Next example loads only "ETH" perpetual prices already stored in the data directory on an hourly frequency (default but can be changed). It returns a pandas DataFrame with columns 'datetime' in UTC.

In [ ]:
#|eval: false
manager = HyperliquidPerpManager(ticker='ETH',update=False, save=False,verbose=True)
manager.get_data()

Processing 1 perpetual tokens...
Loaded 6250 rows for ETH from ../data/hyperliquid/perp/ETH_1h.parquet


,datetime,open,high,low,close,volume,coin
0,2025-03-19 15:00:00+00:00,2030.7,2056.7,2026.0,2048.7,30206.2723,ETH
1,2025-03-19 16:00:00+00:00,2048.7,2049.2,2035.3,2047.4,22198.4803,ETH
2,2025-03-19 17:00:00+00:00,2047.4,2048.9,2014.3,2026.1,35909.8007,ETH
3,2025-03-19 18:00:00+00:00,2026.1,2059.9,1998.0,2045.0,84148.3569,ETH
4,2025-03-19 19:00:00+00:00,2045.1,2052.1,2020.1,2029.7,38921.5619,ETH
...,...,...,...,...,...,...,...
6245,2025-12-04 20:00:00+00:00,3117.9,3143.8,3104.4,3140.5,15553.2974,ETH
6246,2025-12-04 21:00:00+00:00,3140.2,3150.0,3117.2,3121.8,13448.5987,ETH
6247,2025-12-04 22:00:00+00:00,3122.1,3147.2,3121.5,3142.4,11888.6633,ETH
6248,2025-12-04 23:00:00+00:00,3142.6,3142.6,3127.9,3131.7,4731.0506,ETH


If you want to make sure you are featching the most recent data, use the `update=True` parameter. In that case one needs to pass the `info` parameter with your API key to retrieve real-time data which is setup beforehand. The next example demonstrates how to use this:

In [ ]:
#|eval: false
address, info, exchange = setup(base_url=constants.MAINNET_API_URL, skip_ws=True)
manager = HyperliquidPerpManager(ticker='ETH',update=True, save=False, verbose=True,info=info)
manager.get_data()

Processing 1 perpetual tokens...
Loaded 6250 rows for ETH from ../data/hyperliquid/perp/ETH_1h.parquet
Updating ETH from 2025-12-04T00:00:00Z
Retrieved 25 new rows for ETH
Added 0 new rows for ETH


,datetime,open,high,low,close,volume,coin
0,2025-03-19 15:00:00,2030.7,2056.7,2026.0,2048.7,30206.2723,ETH
1,2025-03-19 16:00:00,2048.7,2049.2,2035.3,2047.4,22198.4803,ETH
2,2025-03-19 17:00:00,2047.4,2048.9,2014.3,2026.1,35909.8007,ETH
3,2025-03-19 18:00:00,2026.1,2059.9,1998.0,2045.0,84148.3569,ETH
4,2025-03-19 19:00:00,2045.1,2052.1,2020.1,2029.7,38921.5619,ETH
...,...,...,...,...,...,...,...
6245,2025-12-04 20:00:00,3117.9,3143.8,3104.4,3140.5,15553.2974,ETH
6246,2025-12-04 21:00:00,3140.2,3150.0,3117.2,3121.8,13448.5987,ETH
6247,2025-12-04 22:00:00,3122.1,3147.2,3121.5,3142.4,11888.6633,ETH
6248,2025-12-04 23:00:00,3142.6,3142.6,3127.9,3131.7,4731.0506,ETH


The same goes for the other data items. Such as spot prices and funding rates. Next example shows the spot prices for "ETH" on an hourly frequency and updates it:

In [ ]:
#|eval: false
manager = HyperliquidSpotManager(ticker='ETH',update=True, save=False, verbose=True,info=info)
manager.get_data()

Loaded 5849 rows for ETH from ../data/hyperliquid/spot/ETH_USDC_1h.parquet
Updating ETH/USDC from 2025-12-04T00:00:00Z
Retrieved 25 new rows for ETH/USDC
Added 0 new rows for ETH/USDC


,datetime,open,high,low,close,volume,coin
0,2025-04-05 08:00:00,1812.3,1813.9,1809.7,1812.7,1.7000,ETH
1,2025-04-05 09:00:00,1814.0,1821.0,1814.0,1819.9,19.6808,ETH
2,2025-04-05 10:00:00,1820.2,1820.2,1816.2,1817.4,12.5505,ETH
3,2025-04-05 11:00:00,1818.4,1818.5,1806.6,1806.8,63.9187,ETH
4,2025-04-05 12:00:00,1806.8,1806.9,1793.0,1796.5,45.5064,ETH
...,...,...,...,...,...,...,...
5844,2025-12-04 20:00:00,3120.6,3145.6,3106.6,3143.1,202.4952,ETH
5845,2025-12-04 21:00:00,3141.0,3154.2,3120.0,3123.9,339.5227,ETH
5846,2025-12-04 22:00:00,3125.1,3148.6,3124.4,3144.5,44.5737,ETH
5847,2025-12-04 23:00:00,3141.9,3144.7,3130.9,3133.9,54.7757,ETH


If you want to update the funding rate, do the following:

In [ ]:
#|eval: false
manager = HyperliquidFundingManager(ticker='ETH',update=True, save=False, verbose=True,info=info)
manager.get_data()

Loaded 738 rows for ETH from ../data/hyperliquid/funding/ETH.parquet
Updating ETH from 2025-11-18T11:00:00Z
Retrieved 398 new rows for ETH
Added 373 new rows for ETH


,datetime,funding_rate,premium,coin,fund_calc
0,2025-10-03 21:00:00,0.000013,0.000185,ETH,0.000013
1,2025-10-03 22:00:00,0.000013,0.000111,ETH,0.000013
2,2025-10-03 23:00:00,0.000013,0.000141,ETH,0.000013
3,2025-10-04 00:00:00,0.000013,0.000113,ETH,0.000013
4,2025-10-04 01:00:00,0.000013,0.000081,ETH,0.000013
...,...,...,...,...,...
1106,2025-12-04 20:00:00,0.000009,-0.000429,ETH,0.000009
1107,2025-12-04 21:00:00,0.000010,-0.000422,ETH,0.000010
1108,2025-12-04 22:00:00,0.000009,-0.000427,ETH,0.000009
1109,2025-12-04 23:00:00,0.000007,-0.000442,ETH,0.000007
